### PR TITLE
[LLVM][AsmParser] Make error reporting of lexer errors more precise

### PR DESCRIPTION
**Pull request by jurahul**
When the lexer hits an error during assembly parsing, it just logs the error in the `ErrorMsg` object, and it's possible that error gets overwritten later in by the parser with a more generic error message. This makes some errors reported by the LLVM's asm parser less precise.

Address this by not having the parser overwrite the message logged by the lexer by assigning error messages generated by the lexer higher "priority" than those generated by parser and overwriting the error message only if its same or higher priority.

Update several Assembler unit test to now check the more precise error messaged reported by the LLVM's AsmParser.

### CODE REVIEW

File: llvm/test/Assembler/allockind-missing.ll [hunk 1]
```diff
@@ -1,4 +1,4 @@
-; RUN: not llvm-as %s -o /dev/null 2>&1 | FileCheck %s
+; RUN: not llvm-as --disable-output %s  2>&1 | FileCheck -DFILE=%s %s
 
+; CHECK: [[FILE]]:[[@LINE+1]]:30: error: expected allockind value
```
Comment (jurahul) [inline llvm/test/Assembler/allockind-missing.ll:3]: No real change here, just adding file name as well to the check

File: llvm/test/Assembler/invalid-name.ll
Comment (jurahul) [inline llvm/test/Assembler/invalid-name.ll:1]: These 2 files have strings with zero in them. for the LLVM value name. Both now show "error: Null bytes are not allowed in names" as opposed to "expected instruction opcode" which is less precise.

**Review by arsenm — COMMENTED** [review]
Does this help with #29043?

**Comment by jurahul** [comment]
> Does this help with #29043?

Unfortunately, it does not. The <2 x i32> is a set of valid tokens, so this PR which makes the lexer errors sticky won't address that issue.

File: llvm/include/llvm/AsmParser/LLLexer.h [hunk 2]
```diff
@@ -66,8 +79,13 @@ namespace llvm {
       IgnoreColonInIdentifiers = val;
     }
 
-    bool Error(LocTy ErrorLoc, const Twine &Msg) const;
-    bool Error(const Twine &Msg) const { return Error(getLoc(), Msg); }
+    // This returns true as a convienence for the parser functions that return
```
Comment (arsenm) [inline llvm/include/llvm/AsmParser/LLLexer.h:82]:
```suggestion
    // This returns true as a convenience for the parser functions that return
```

File: llvm/test/Assembler/invalid-name.ll
Comment (arsenm) [inline llvm/test/Assembler/invalid-name.ll:1]: Why does GitHub think this is an undisplayable binary file

File: llvm/test/Assembler/invalid-name2.ll
Comment (arsenm) [inline llvm/test/Assembler/invalid-name2.ll:1]: Why does GitHub think this is an undisplayable binary file

File: llvm/test/Assembler/invalid-name.ll
Comment (jurahul) [reply llvm/test/Assembler/invalid-name.ll:1]: @arsenm ^^

Comment (jurahul) [reply llvm/test/Assembler/invalid-name.ll:1]: Please see my comment above.

File: llvm/test/Assembler/invalid-name2.ll
Comment (jurahul) [reply llvm/test/Assembler/invalid-name2.ll:1]: Please see my comment above.

File: llvm/test/Assembler/invalid-name.ll
Comment (jurahul) [reply llvm/test/Assembler/invalid-name.ll:1]: These 2 files have strings with zero in them for the LLVM value name. Both now show "error: Null bytes are not allowed in names" as opposed to "expected instruction opcode" which is less precise.

File: llvm/include/llvm/AsmParser/LLLexer.h (hunk 2 above)
Comment (jurahul) [reply llvm/include/llvm/AsmParser/LLLexer.h]: Done

**Review by arsenm — APPROVED** [review]

**Pull request merged**